go/common/workerpool/workerpool.go

```diff
@@ -130,23 +203,37 @@ func (p *Pool) worker() {
 			if job.terminate {
 				return
 			}
-			job.job()
+			err := job.job()
 			close(job.completeCh)
+
+			// Submit backoff feedback.
+			if err == nil {
+				p.backoff.Success()
+			} else {
+				p.backoff.Failure()
+			}
 		}
 	}
 }
 
 // New creates and returns a new worker pool with one worker goroutine.
-func New(name string) *Pool {
+func New(name string, backoffCfg *BackoffConfig) *Pool {
```

Comment:
As backoff is optional, I would remove it from the constructor, add a method `WithBackoff`, remove `noBackoff` implementation and rather check if `backoff != nil`.

I guess it is up to debate if the workerpool should be responsible for backoffs or the jobs. Currently, mixing jobs A and B, where A always fails and needs a backoff and B always succeeds, would force B to wait, and the latter would always reset the timer.

Reply:
I thought about it, but the `WithBackoff` version has a downside that the field now needs to be protected by a lock, because it won't be ensured that it is configured before jobs are submitted. The nicest approach would likely be something akin to an option builder pattern, but seems overkill for this single option.
I think passing an optional option/config struct is quite a common pattern in go.

>remove noBackoff implementation and rather check if backoff != nil. 

Any reasons why preferring the `backoff != nil` variant? I had both implemented, seemed nicer to avoid the `backoff != nil` checks. The only benefit I see is that it avoids the need for the internal `backoff` interface. Don't have a strong opinion about either though.

> I guess it is up to debate if the workerpool should be responsible for backoffs or the jobs. Currently, mixing jobs A and B, where A always fails and needs a backoff and B always succeeds, would force B to wait, and the latter would always reset the timer

For cases where jobs are not mixed, having this in the workerpool avoids duplicating it, so it kinda makes sense i think? 

> As currently the pool is responsible, the lock is actually not needed.

Which lock? The one in `expBackoff`? It's needed.

Reply:
> I think passing an optional option/config struct is quite a common pattern in go.

That's true. Another option would also be to create 2 constructors. And we should probably not add the config to the constructor, but an interface instead. The backoff constructor should take the config.

> Any reasons why preferring the backoff != nil variant? I

To avoid creating a new timer with zero duration and giving garbage collector less work.

> For cases where jobs are not mixed, having this in the workerpool avoids duplicating it, so it kinda makes sense i think?

Yeah, its probably fine. Was just saying because the pool accepts arbitrary jobs, so they could be mixed.

> Which lock? The one in expBackoff is needed.

I would say not, as `Success`, `Failure` and `Timeout` are called only in `worker` method.

Reply:
> To avoid creating a new timer with zero duration and giving garbage collector less work.

:+1: switched already in any case.

> are called only in worker method.

There can be multiple workers https://github.com/oasisprotocol/oasis-core/blob/457ee23f9a772e78a900d2fd1875132ceb1d978f/go/common/workerpool/workerpool.go#L62-L64